### PR TITLE
feat: add configurable timeout for unsent messages (#944)

### PR DESCRIPTION
**Pull request by jonas-jonas**
This PR adds the option to configure a TTL (Time-To-Live) for messages that can not be delivered. After the TTL has passed since the message's creation and it was not delivered, the message is `Abandoned` and the courier will not try to deliver the message again. This is done by setting its `status` to `MessageStatusAbandoned` in the database. The courier will only try to deliver messages with the status `MessageStatusQueued`. 

## Related issue(s)

This PR resolves https://github.com/ory/kratos/issues/944

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [x] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security. vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [x] I have added or changed [the documentation](docs/docs).

### CODE REVIEW

**Comment by jonas-jonas** [comment]
Hi @Benehiko

thanks for the detailed answer. I have added a few tests to the relevant classes. I also added the key to the `config.schema.json` file. 

This should be everything and the PR is ready-for-review.

File: embedx/config.schema.json [hunk 1]
```diff
@@ -1394,6 +1394,17 @@
...
+          "type": "string",
+          "pattern": "^([0-9]+(ns|us|ms|s|m|h))+$",
+          "default": "1h",
+          "examples": [
+            "1h",
+            "1m",
+            "1s"
+          ]
+        },
         "smtp": {
```
Comment (Benehiko) [inline embedx/config.schema.json:1408]: Nice! :)

**Review by Benehiko — CHANGES_REQUESTED** [review]
This looks very thorough! Very nicely done :)

I have a few comments to improve this PR:
- We need a check in the persistence layer to not query abandoned projects. Somewhere [here](https://github.com/ory/kratos/blob/master/persistence/sql/persister_courier.go#L27-L39). This also means a corresponding test to make sure the `courier.NextMessages` function does not return abandoned messages.
- We need configuration tests [here](https://github.com/ory/kratos/blob/master/driver/config/config_test.go) and [here](https://github.com/ory/kratos/tree/master/test/schema). 

Other than that I think this is quite close to being merged

**Comment by jonas-jonas** [comment]
> We need a check in the persistence layer to not query abandoned projects. Somew[here](https://github.com/ory/kratos/blob/master/persistence/sql/persister_courier.go#L27-L39) here. This also means a corresponding test to make sure the courier.NextMessages function does not return abandoned messages.

That method already only queries messages that have the `status=MessageStatusQueued`. I think that should exclude messages that have the `status= MessageStatusAbandoned` (or any other `MessageStatus` for that matter). This is also tested [here](https://github.com/ory/kratos/blob/b1f270c12e7193f3d2abb65947ed92c1906ee69c/courier/test/persistence.go#L93-L95) by manually setting the `status` of a message to Abandoned and then checking the result to be empty. Am I missing something here?


>We need configuration tests [here](https://github.com/ory/kratos/blob/master/driver/config/config_test.go) and [here](https://github.com/ory/kratos/tree/master/test/schema).

I will add the tests later today.

**Comment by Benehiko** [comment]
> > We need a check in the persistence layer to not query abandoned projects. Somew[here](https://github.com/ory/kratos/blob/master/persistence/sql/persister_courier.go#L27-L39) here. This also means a corresponding test to make sure the courier.NextMessages function does not return abandoned messages.
> 
> That method already only queries messages that have the `status=MessageStatusQueued`. I think that should exclude messages that have the `status= MessageStatusAbandoned` (or any other `MessageStatus` for that matter). This is also tested [here](https://github.com/ory/kratos/blob/b1f270c12e7193f3d2abb65947ed92c1906ee69c/courier/test/persistence.go#L93-L95) by manually setting the `status` of a message to Abandoned and then checking the result to be empty. Am I missing something here?
> 
> > We need configuration tests [here](https://github.com/ory/kratos/blob/master/driver/config/config_test.go) and [here](https://github.com/ory/kratos/tree/master/test/schema).
> 
> I will add the tests later today.

Ah yes, sorry about that. You are correct :)

**Comment by jonas-jonas** [comment]
I added the config tests. This should be everything.

File: courier/test/persistence.go [hunk 2]
```diff
@@ -89,6 +89,10 @@ func TestPersister(ctx context.Context, newNetworkUnlessExisting NetworkWrapper,
 			require.NoError(t, p.SetMessageStatus(ctx, messages[0].ID, courier.MessageStatusSent))
 			_, err = p.NextMessages(ctx, 1)
 			require.ErrorIs(t, err, courier.ErrQueueEmpty)
+
+			require.NoError(t, p.SetMessageStatus(ctx, messages[0].ID, courier.MessageStatusAbandoned))
```
Comment (zepatrik) [inline courier/test/persistence.go:93]: Ah, I see. The persister already only gets pending ones.

**Review by zepatrik — APPROVED** [review]
Nice, looks good to me :+1: 
Waiting for @Benehiko to confirm.

**Review by Benehiko — APPROVED** [review]
LGTM :)

Really nicely done!

**Comment by vinckr** [comment]
Hello @jonas-jonas 
Congrats on merging your first PR in Ory 🎉 !
Your contribution will soon be helping secure millions of identities around the globe 🌏. 
As a small token of appreciation we send all our first time contributors a gift package to welcome them to the community. 
[Please drop me an email and I will forward you the form to claim your Ory swag!](mailto:vincent@ory.sh)

**Pull request merged**